crates/runtime/src/instance/allocator/pooling/uffd.rs

```diff
@@ -0,0 +1,833 @@
+            });
+        }
+
+        None
+    }
+}
+
+unsafe fn wake_guard_page_access(
+    uffd: &Uffd,
+    page_addr: *const u8,
+    len: usize,
+) -> Result<(), String> {
+    // Set the page to NONE to induce a SIGSEV for the access on the next retry
+    region::protect(page_addr, len, region::Protection::NONE)
+        .map_err(|e| format!("failed to change guard page protection: {}", e))?;
+
+    uffd.wake(page_addr as _, len).map_err(|e| {
+        format!(
+            "failed to wake page at {:p} with length {}: {}",
+            page_addr, len, e
+        )
+    })?;
+
+    Ok(())
+}
+
+unsafe fn initialize_wasm_page(
+    uffd: &Uffd,
+    instance: &Instance,
+    page_addr: *const u8,
+    memory_index: usize,
+    page_index: usize,
+) -> Result<(), String> {
+    if let Some(MemoryInitialization::Paged { page_size, map }) =
+        &instance.module.memory_initialization
+    {
+        let memory_index = DefinedMemoryIndex::new(memory_index);
+        let memory = instance.memory(memory_index);
+        let pages = &map[memory_index];
+        debug_assert_eq!(WASM_PAGE_SIZE % page_size, 0);
+
+        let count = WASM_PAGE_SIZE / page_size;
+        let start = page_index * count;
+
+        for i in start..start + count {
+            let dst = memory.base.add(i * page_size);
+
+            match pages.get(i) {
+                Some(Some(data)) => {
+                    log::trace!(
+                        "copying page initialization data from {:p} to {:p} with length {}",
+                        data,
+                        dst,
+                        page_size
+                    );
+
+                    // Copy the page data without waking
+                    uffd.copy(data.as_ptr() as _, dst as _, *page_size, false)
+                        .map_err(|e| {
+                            format!(
+                                "failed to copy page from {:p} to {:p} with length {}: {}",
+                                data, dst, page_size, e
+                            )
+                        })?;
+                }
+                _ => {
+                    log::trace!("zeroing page at {:p} with length {}", dst, page_size);
+
+                    // No data, zero the page without waking
+                    uffd.zeropage(dst as _, *page_size, false).map_err(|e| {
+                        format!(
+                            "failed to zero page at {:p} with length {}: {}",
+                            dst, page_size, e
+                        )
+                    })?;
+                }
+            }
+        }
+
+        // Finally wake the entire wasm page
+        uffd.wake(page_addr as _, WASM_PAGE_SIZE).map_err(|e| {
+            format!(
+                "failed to wake page at {:p} with length {}: {}",
+                page_addr, WASM_PAGE_SIZE, e
+            )
+        })
+    } else {
+        log::trace!(
+            "initialization data is not paged; zeroing Wasm page at {:p}",
```

Comment:
I'm having a bit of trouble imagining where this case is hit. Is this used for things like `Memory::new` explicit constructors?

It seems like this also gets triggered perhaps when uffd is used but pages are not initialized with the paging above?

I feel like ideally we would assert that this case was never hit, but I'm not sure if that makes sense within the realm of performance that uffd is targeting.

Reply:
So this happens when the data segments either initialize imported memories or use a global base (this is treated as "segmented" initialization instead of "paged").

For the bulk memory proposal, initializing an imported memory should have visible side effects if the module fails to instantiate, so we can't delay-initialize the pages.  Thus we handle the fault by zeroing the page; the fault is actually caused by Wasmtime during module instantiation, so the next thing that's happening when we wake the thread is that it's copying the segment data into the page. 

I just realized we *can* actually delay-initialize when a global base is used (for defined memories) because it's required to be an immutable global, so the value cannot change from the time the module instantiates to the time the page is accessed.  I'll fix that.

Reply:
Ok cool makes sense, could you add some comments to that effect here of when this case will be hit?